curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,418 @@
+}
+
+virtual MetaStatusCode S3CompactWorkQueueImpl::updateInode(
+    CopysetNode* copysetNode, const PartitionInfo& pinfo, const Inode& inode) {
+    UpdateInodeRequest request;
+    request.set_poolid(pinfo.poolid());
+    request.set_copysetId(pinfo.copysetId());
+    request.set_partitionid(pinfo.partitionid());
+    request.set_fsid(pinfo.fsid());
+    request.set_inodeid(inode.inodeid());
+    request.set_s3chunkinfomap(inode.s3chunkinfomap());
+    UpdateInodeResponse response;
+    S3CompactWorkQueueImpl::updateInodeClosure done;
+    auto updateInodeOp = new UpdateInodeOperator(copysetNode, nullptr, &request,
+                                                 &respnse, &done);
+    updateInodeOp->Propose();
+    done.waitRunned();
+    return response.statusCode;
+}
+
+int S3CompactWorkQueueImpl::writeFullChunk(
+    const std::string& fullChunk, uint64_t inodeId, uint64_t blockSize,
+    uint64_t newChunkid, uint64_t newCompaction,
+    std::vector<std::string>* objsAdded) {
+    uint64_t chunkLen = fullChunk.length();
+    for (uint64_t index = 0; index * blockSize < chunkLen; index += 1) {
+        std::string objName =
+            std::to_string(newChunkid) + "_" + std::to_string(newCompaction) +
+            "_" + std::to_string(index) + "_" + std::to_string(inodeId);
+        const Aws::String aws_key(objName.c_str(), objName.size());
+        int ret;
+        if (chunkLen > (index + 1) * blockSize) {
+            ret = s3Adapter_->PutObject(
+                aws_key, fullChunk.substr(index * blockSize, blockSize));
+        } else {
+            // last block
+            ret = s3Adapter_->PutObject(
+                aws_key, fullChunk.substr(index * blockSize,
+                                          chunkLen - index * blockSize));
+        }
+        if (ret != 0) {
+            return ret;
+        } else {
+            objsAdded->emplace_back(std::move(objName));
+        }
+    }
+    return 0;
+}
+
+void S3CompactWorkQueueImpl::CompactChunks(
+    std::shared_ptr<InodeStorage> inodeStorage, const InodeKey& inodeKey,
+    const PartitionInfo& pinfo, const CopysetNode* copysetNode) {
+    uint64_t blockSize;
+    LOG_IF(FATAL, !conf_->GetUInt64Value("s3.blocksize", &blockSize));
+
+    // inode exist?
+    Inode inode;
+    MetaStatusCode ret = inodeStorage->Get(inodeKey, &inode);
+    if (ret != MetaStatusCode::OK) {
+        LOG(ERROR) << "GetInode fail, inodeKey = " << inodeKey.fsId << ","
+                   << inodeKey.inodeId
+                   << ", ret = " << MetaStatusCode_Name(ret);
+        return;
+    }
+
+    // deleted?
+    if (inode.nlink() == 0) {
+        return;
+    }
+
+    // s3chunkinfomap size 0?
+    if (inode.s3chunkinfomap().size() == 0) {
+        return;
+    }
+    uint64_t inodeId = inode.inodeid();
+    uint64_t inodeLen = inode.length();
+
+    // need compact?
+    const auto origMap(inode.s3chunkinfomap());
+    std::vector<uint64_t> needCompact(std::move(getNeedCompact(origMap)));
```

Comment:
no need to add an extra std::move, return value is rvalue by default.

Reply:
fixed.